README


This archive contains the distribution Sub-Exporter,
version 0.986:

  a sophisticated exporter for custom-built routines

This software is copyright (c) 2007 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Sub-Exporter

0.986     2013-06-14 18:45:45 America/New_York
          typo fixes in docs (thanks, David Steinbrunner!)

0.985     2013-02-20 19:02:30 America/New_York
          documentation fixes (thanks, George Hartzell)

0.984     2012-06-05 07:59:40 America/New_York
          documentation fixes (thanks, GitHub user "everybody")

0.983     2011-01-24
          documentation fixes (thanks, Karen Etheridge and Luc St-Louis!)

0.982     2009-01-16
          add metadata for repo

0.981     2008-10-24
          finally fix very occasional hash ordering issue in tests
          fix typo in SYNOPSIS (thanks, Florian!)

0.980     2008-09-14
          fix inadvertant futzing with group generator args
            https://rt.cpan.org/Ticket/Display.html?id=38885
            thanks, trendele!

0.979     2008-04-29
          add INIT collector
          declare reservation of all CAPS collectors
          clarify documentation of -setup after report by GAISSMAI

0.978     2007-11-19
          improve documentation of new installer/generator options
          deprecate calling "installer" the "exporter"
          WARNING: "exporter" OPTION WILL BE REMOVED AFTER 2008-06-01
          major refactoring of the core generation/installation code
          tentative interface documentation for replacing it!

0.976     2007-08-30
          fixed merge_col, which was not updated to work with \name generators
          collector hooks can now alter @_ to replace the value to be collected
          clarify args passed to generator in Tutorial; thanks MARKSTOS

          added commented-out name_map to Sub::Exporter::Util; future feature?

0.975     2007-07-04
          update Tutorial to show (preferred) \'name' style for generators
          changed "standard" name of curry_class to curry_method
          added curry_chain
          added Sub::Exporter::Cookbook

0.974     2007-04-22
          fix a bug: would try to export routines that didn't exist
          in the exporting package; this caused Sub::Install to give the
          unhelpful message "argument 'code' is not optional"

0.973     2007-02-02
          document changes made in 0.972
          minor code changes for readability

0.972     2006-12-05
          allow exporter config to provide name (via string ref) of generator
          for groups and exports
          similarly allow a string ref for a method name for a collector hook
          remove some pointless conditions

0.971     2006-11-06
          minor documentation clarification
          add Perl::Critic tests (disabled by default)

0.970     2006-06-27
          defaults populate before collectors collect, now
          default group's value is undef by default, not 1
          mixin_exporter can now export into objects, creating virtual classes
          
0.966     2006-06-17
          correct documentation of collector hook args
          simplify internal use of setup_exporter
          clean up documentation in ::Util

0.965     2006-06-05
          curry_class now allows the export to curry a differently-named method

0.961     2006-06-05
          Data::OptList is now in its own dist; updated to use it

0.960     2006-05-31
          added into and into_config to config
          100% test coverage... almost!
          fix bug that prevented validation of opt lists with must_be=class

0.954     2006-05-11
          tweaks to Data::OptList, moving toward its own dist: now it exports
          expand_opt_list is now opt_list_as_hash

0.953     2006-05-10
          require Params::Util for craftier opt list validation
          use reinstall, rather than install, to avoid warnings on redef

0.952     2006-04-30
          add missing file to manifest

0.951     2006-04-30
          fix util-mixin.t to skip if prereqs are missing
          various changes to improve blessed/weird generators
          (thanks to Yuval Kogman for pointing problems out)

0.95      2006-04-26
          break out Data::OptList for future disting
          remove an "optimization" that broke expand_opt_list
          improve detection of group generators
          improve data passed to hooks (if you relied on the guts, you'll break)
          the ::Util module

0.93      2006-03-26
          internal refactoring
          add more arguments to collector hook calls

0.92      2006-03-16
          FIX BUG in nested imports: when importing groups A and B, and group B
          includes group A, the nested group would be ignored, even though it
          was not recursing

          allow 'into_level' parameter to setup_exporter
          rewrite collection collector to be more efficient
          rewrite opt list handlers to be more efficient
          restate some code to improve clarity and coverage (now 100%)
          better diagnostic messages

0.91      2006-03-16
          added "import elsewhere" option to generated exporter (thanks
          chansen!)

0.90      2006-03-11
          first public release


lib/Sub/Exporter/Util.pm
use strict;
use warnings;
package Sub::Exporter::Util;
BEGIN {
  $Sub::Exporter::Util::VERSION = '0.983'; # TRIAL
}
# ABSTRACT: utilities to make Sub::Exporter easier

use Data::OptList ();
use Params::Util ();


sub curry_method {
  my $override_name = shift;
  sub {
    my ($class, $name) = @_;
    $name = $override_name if defined $override_name;
    sub { $class->$name(@_); };
  }
}

BEGIN { *curry_class = \&curry_method; }


sub curry_chain {
  # In the future, we can make \%arg an optional prepend, like the "special"
  # args to the default Sub::Exporter-generated import routine.
  my (@opt_list) = @_;

  my $pairs = Data::OptList::mkopt(\@opt_list, 'args', 'ARRAY');

  sub {
    my ($class) = @_;

    sub {
      my $next = $class;

      for my $i (0 .. $#$pairs) {
        my $pair = $pairs->[ $i ];
        
        unless (Params::Util::_INVOCANT($next)) { ## no critic Private
          my $str = defined $next ? "'$next'" : 'undef';
          Carp::croak("can't call $pair->[0] on non-invocant $str")
        }

        my ($method, $args) = @$pair;

        if ($i == $#$pairs) {
          return $next->$method($args ? @$args : ());
        } else {
          $next = $next->$method($args ? @$args : ());
        }
      }
    };
  }
}

# =head2 name_map
# 
# This utility returns an list to be used in specify export generators.  For
# example, the following:
# 
#   exports => {
#     name_map(
#       '_?_gen'  => [ qw(fee fie) ],
#       '_make_?' => [ qw(foo bar) ],
#     ),
#   }
# 
# is equivalent to:
# 
#   exports => {
#     name_map(
#       fee => \'_fee_gen',
#       fie => \'_fie_gen',
#       foo => \'_make_foo',
#       bar => \'_make_bar',
#     ),
#   }
# 
# This can save a lot of typing, when providing many exports with similarly-named
# generators.
# 
# =cut
# 
# sub name_map {
#   my (%groups) = @_;
# 
#   my %map;
# 
#   while (my ($template, $names) = each %groups) {
#     for my $name (@$names) {
#       (my $export = $template) =~ s/\?/$name/
#         or Carp::croak 'no ? found in name_map template';
# 
#       $map{ $name } = \$export;
#     }
#   }
# 
#   return %map;
# }


sub merge_col {
  my (%groups) = @_;

  my %merged;

  while (my ($default_name, $group) = each %groups) {
    while (my ($export_name, $gen) = each %$group) {
      $merged{$export_name} = sub {
        my ($class, $name, $arg, $col) = @_;

        my $merged_arg = exists $col->{$default_name}
                       ? { %{ $col->{$default_name} }, %$arg }
                       : $arg;

        if (Params::Util::_CODELIKE($gen)) { ## no critic Private
          $gen->($class, $name, $merged_arg, $col);
        } else {
          $class->$$gen($name, $merged_arg, $col);
        }
      }
    }
  }

  return %merged;
}


sub __mixin_class_for {
  my ($class, $mix_into) = @_;
  require Package::Generator;
  my $mixin_class = Package::Generator->new_package({
    base => "$class\:\:__mixin__",
  });

  ## no critic (ProhibitNoStrict)
  no strict 'refs';
  if (ref $mix_into) {
    unshift @{"$mixin_class" . "::ISA"}, ref $mix_into;
  } else {
    unshift @{"$mix_into" . "::ISA"}, $mixin_class;
  }
  return $mixin_class;
}

sub mixin_installer {
  sub {
    my ($arg, $to_export) = @_;

    my $mixin_class = __mixin_class_for($arg->{class}, $arg->{into});
    bless $arg->{into} => $mixin_class if ref $arg->{into};

    Sub::Exporter::default_installer(
      { %$arg, into => $mixin_class },
      $to_export,
    );
  };
}

sub mixin_exporter {
  Carp::cluck "mixin_exporter is deprecated; use mixin_installer instead; it behaves identically";
  return mixin_installer;
}


sub like {
  sub {
    my ($value, $arg) = @_;
    Carp::croak "no regex supplied to regex group generator" unless $value;

    # Oh, qr//, how you bother me!  See the p5p thread from around now about
    # fixing this problem... too bad it won't help me. -- rjbs, 2006-04-25
    my @values = eval { $value->isa('Regexp') } ? ($value, undef)
               :                                  @$value;

    while (my ($re, $opt) = splice @values, 0, 2) {
      Carp::croak "given pattern for regex group generater is not a Regexp"
        unless eval { $re->isa('Regexp') };
      my @exports  = keys %{ $arg->{config}->{exports} };
      my @matching = grep { $_ =~ $re } @exports;

      my %merge = $opt ? %$opt : ();
      my $prefix = (delete $merge{-prefix}) || '';
      my $suffix = (delete $merge{-suffix}) || '';

      for my $name (@matching) {
        my $as = $prefix . $name . $suffix;
        push @{ $arg->{import_args} }, [ $name => { %merge, -as => $as } ];
      }
    }

    1;
  }
}

use Sub::Exporter -setup => {
  exports => [ qw(
    like
    name_map
    merge_col
    curry_method curry_class
    curry_chain
    mixin_installer mixin_exporter
  ) ]
};

1;

__END__
=pod

=head1 NAME

Sub::Exporter::Util - utilities to make Sub::Exporter easier

=head1 VERSION

version 0.983

=head1 DESCRIPTION

This module provides a number of utility functions for performing common or
useful operations when setting up a Sub::Exporter configuration.  All of the
utilites may be exported, but none are by default.

=head1 THE UTILITIES

=head2 curry_method

  exports => {
    some_method => curry_method,
  }

This utility returns a generator which will produce an invocant-curried version
of a method.  In other words, it will export a method call with the exporting
class built in as the invocant.

A module importing the code some the above example might do this:

  use Some::Module qw(some_method);

  my $x = some_method;

This would be equivalent to:

  use Some::Module;

  my $x = Some::Module->some_method;

If Some::Module is subclassed and the subclass's import method is called to
import C<some_method>, the subclass will be curried in as the invocant.

If an argument is provided for C<curry_method> it is used as the name of the
curried method to export.  This means you could export a Widget constructor
like this:

  exports => { widget => curry_method('new') }

This utility may also be called as C<curry_class>, for backwards compatibility.

=head2 curry_chain

C<curry_chain> behaves like C<L</curry_method>>, but is meant for generating
exports that will call several methods in succession.

  exports => {
    reticulate => curry_chain([
      new => gather_data => analyze => [ detail => 100 ] => results
    ]),
  }

If imported from Spliner, calling the C<reticulate> routine will be equivalent
to:

  Splinter->new->gather_data->analyze(detail => 100)->results;

If any method returns something on which methods may not be called, the routine
croaks.

The arguments to C<curry_chain> form an optlist.  The names are methods to be
called and the arguments, if given, are arrayrefs to be dereferenced and passed
as arguments to those methods.  C<curry_chain> returns a generator like those
expected by Sub::Exporter.

B<Achtung!> at present, there is no way to pass arguments from the generated
routine to the method calls.  This will probably be solved in future revisions
by allowing the opt list's values to be subroutines that will be called with
the generated routine's stack.

=head2 merge_col

  exports => {
    merge_col(defaults => {
      twiddle => \'_twiddle_gen',
      tweak   => \&_tweak_gen,
    }),
  }

This utility wraps the given generator in one that will merge the named
collection into its args before calling it.  This means that you can support a
"default" collector in multipe exports without writing the code each time.

You can specify as many pairs of collection names and generators as you like.

=head2 mixin_installer

  use Sub::Exporter -setup => {
    installer => Sub::Exporter::Util::mixin_installer,
    exports   => [ qw(foo bar baz) ],
  };

This utility returns an installer that will install into a superclass and
adjust the ISA importing class to include the newly generated superclass.

If the target of importing is an object, the hierarchy is reversed: the new
class will be ISA the object's class, and the object will be reblessed.

B<Prerequisites>: This utility requires that Package::Generator be installed.

=head2 like

It's a collector that adds imports for anything like given regex.

If you provide this configuration:

  exports    => [ qw(igrep imap islurp exhausted) ],
  collectors => { -like => Sub::Exporter::Util::like },

A user may import from your module like this:

  use Your::Iterator -like => qr/^i/; # imports igre, imap, islurp

or

  use Your::Iterator -like => [ qr/^i/ => { -prefix => 'your_' } ];

The group-like prefix and suffix arguments are respected; other arguments are
passed on to the generators for matching exports.

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "a sophisticated exporter for custom-built routines",
  "AUTHOR" => "Ricardo Signes <rjbs\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Sub-Exporter",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Sub::Exporter",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::OptList" => "0.100",
    "Params::Util" => "0.14",
    "Sub::Install" => "0.92",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Test::More" => "0.96",
    "base" => 0,
    "overload" => 0,
    "subs" => 0
  },
  "VERSION" => "0.986",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



